deltacat/storage/model/metafile.py

```diff
@@ -293,115 +634,211 @@ def current(
-                        f"transaction ID {txn_id} failed. Metafile state at "
-                        f"{mci.path} is deleted."
+                        f"Metafile {current_txn_op_type.value} failed "
+                        f"for transaction ID {current_txn_id} failed. "
+                        f"Metafile state at {mri.path} is deleted."
                     )
             # create fails unless the last metafile was deleted
             elif is_create_txn:
                 raise ValueError(
-                    f"Metafile creation for transaction ID {txn_id} failed. "
-                    f"Metafile commit at {mci.path} already exists."
+                    f"Metafile creation for transaction ID {current_txn_id} "
+                    f"failed. Metafile commit at {mri.path} already exists."
                 )
         elif not is_create_txn:
-            # update/deletes fails if the last metafile doesn't exist
+            # update/delete fails if the last metafile doesn't exist
             raise ValueError(
-                f"Metafile {txn_operation_type.value} failed for "
-                f"transaction ID {txn_id} failed. Metafile at {mci.path} "
-                f"doesn't exist."
+                f"Metafile {current_txn_op_type.value} failed for "
+                f"transaction ID {current_txn_id} failed. Metafile at "
+                f"{mri.path} does not exist."
             )
+        mri.revision = mri.revision + 1
+        mri.txn_id = current_txn_id
+        mri.txn_op_type = current_txn_op_type
+        mri.dir_path = revision_dir_path
+        mri.extension = extension
+        return mri
+
+    @staticmethod
+    def check_for_concurrent_txn_conflict(
+        txn_log_dir: str,
+        current_txn_revision_file_path: str,
+        filesystem: pyarrow.fs.FileSystem,
+    ) -> None:
+        """
+        Checks for a concurrent modification conflict between a file commited
+        by the current transaction and another parallel transaction. Raises
+        an exception if a concurrent modification conflict is found.
+
+        :param txn_log_dir: Path to the catalog transaction log.
+        :param current_txn_revision_file_path: Path to a metafile revision
+        written by the current transaction to check for conflicts against.
+        :param filesystem: Filesystem that can read the metafile revision.
+        :raises RuntimeError: if a conflict is found with another transaction.
+        """
+        revision_dir_path = posixpath.dirname(current_txn_revision_file_path)
+        cur_txn_mri = MetafileRevisionInfo.parse(current_txn_revision_file_path)
 
-        mci.next_txn_id = txn_id
-        mci.next_txn_operation_type = txn_operation_type
-        metafile_name = TXN_ID_SEPARATOR.join(
-            [
-                f"{mci.next_revision:020}",
-                txn_operation_type.value,
-                f"{txn_id}{extension}",
-            ]
+        sorted_metafile_paths = MetafileRevisionInfo._sorted_file_paths(
```

Comment:
I think we can improve performance here by writing a file like `_latest.mpk` in each rev directory populated with the latest metafile revision, and then putting each revision file in its own revision number directory (such that this method only needs to list the contents of the latest revision number directory to check for conflicts, and that metafile reads don't need to list revision contents to find the latest revision (they just go directly to reading `_latest.mpk`).